aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -0,0 +1,304 @@
+#include <ATen/Dispatch.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/ReduceOps.h>
+#include <ATen/native/cpu/Reduce.h>
+#include <c10/util/llvmMathExtras.h>
+
+#include <algorithm>
+
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename scalar_t>
+struct LoadImpl {
+  static scalar_t load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    auto *ptr = reinterpret_cast<const scalar_t*>(data + index * stride);
+    return *ptr;
+  }
+};
+
+template <typename scalar_t>
+struct LoadImpl<Vec256<scalar_t>> {
+  static Vec256<scalar_t> load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    auto *ptr = data + index * stride;
+    return Vec256<scalar_t>::loadu(ptr);
+  }
```

Comment:
It's a little weird that these implementations don't already exist in the vec256 headers. Would they logically belong there? Why not?

Reply:
I couldn't see anything similar. It can't find any other kernels that can be templated on scalar or vector types and instead most have specialised code paths for each case.